### ISSUE TITLE
What's the most efficient way to animate a large amount of points with scatterplot layer?

### ISSUE THREAD

**Issue by juscc** [visgl/deck.gl#4947]
I want to animate points with the scatterplot layer for large cities to represent vehicles for an entire day (goal of 30,000 points for 1800 frames for now), what are some approaches I can try? My main question is, is there a better way to utilise the GPU, or better data formats? And is it possible to have the interpolation transition while using data filter extensions on the Scatterplot layer? I don't have any experience with WebGL or OpenGL... but am happy to give it a go if it's relatively straightforward for what I want to achieve. 

This isn't a question about streaming or async data loading or using web workers as that's something separate I can look into. 

Below are the approaches I've tried + more info.

- Currently I'm already supplying attributes directly to the layer as mentioned in the Performance docs https://deck.gl/docs/developer-guide/performance#supply-attributes-directly. 

- The data is lat + long points at each frame / time step

- Scatterplot layer: filtering the data on the CPU/main JS thread and passing each 'frame' into the scatterplot layer as the time ticks. This allowed for smooth animation with interpolation, but if the number of items in the array changed it wouldn't work. Also caused some lag when zooming + moving the map around.

- Scatterplot layer: passing larger sets of data into the scatterplot layer and using the Data filter extension to 'animate'. But this doesn't allow for a smooth animation.

- Trips layer:  I tried using the trips layer but that wasn't efficient enough or scalable for longer/larger amounts of trips. Also didn't handle 'stopped' vehicles well'. 

Thanks in advance!
June

**Comment by kylebarron** [comment]
You could consider the data filter extension https://deck.gl/examples/data-filter-extension/

**Comment by juscc** [comment]
@kylebarron thanks for the reply! Sorry, when I say smooth I meant the interpolation transition between 2 points which I can get if I passed each 'frame' into the Scatterplot as opposed to using the data filter extension. I understand why I can't use interpolation with the data filter extension because it doesn't have the concept of the 2 points being related. But I was wondering if I could get a similar animation with utilising more of the GPU? I might need to look into custom layers but I'm not quite sure where to start with that.

**Comment by Pessimistress** [comment]
ScatterplotLayer + transition is your best bet.

> if the number of items in the array changed it wouldn't work

You need to maintain a list of unique vehicles and always provide data in that order. If a vehicle is invalid at a given timestamp you can hide it using transparent color or zero radius.

> caused some lag when zooming + moving the map around.

This is likely caused by an issue in your implementation. See [Avoid unnecessary shallow change in data prop](https://deck.gl/docs/developer-guide/performance#avoid-unnecessary-shallow-change-in-data-prop).

**Comment by juscc** [comment]
> ScatterplotLayer + transition is your best bet.
Ah okay, thank you, I'll give this another go!

> This is likely caused by an issue in your implementation. See [Avoid unnecessary shallow change in data prop](https://deck.gl/docs/developer-guide/performance#avoid-unnecessary-shallow-change-in-data-prop).

Using React, would I be able to avoid the unnecessary shallow change in data by storing the current frame's data directly using the setState hook? Something like this? I just wrote this up so this is more pseudo code then a working example
`
// object where key is the frame number/timestep e.g. {0: [lat, long, lat, long...], 1: [lat, long, lat long...],...}
const [vehicleBufferData, setVehicleBufferData] = useState({});

// all lat long points for current time only
const [vehicleFrameData, setVehicleFrameData] = useState([]);

//animation loop time
const [time, setTime] = useState(0);

// assume constant stream of data coming in
onReceiveData = (frame, data) => {
	setVehicleBufferData((prev) => {
		return {...prev, [frame]: data}
	});
}


useEffect(() => {
	setVehicleFrameData(vehicleBufferData[time]);
}, [time]);


_renderLayers() {
	return new ScatterplotLayer({
		id: 'vehicle-layer',
		data: vehicleFrameData,
		...
	});
}
`

**Comment by Pessimistress** [comment]
> Currently I'm already supplying attributes directly to the layer as mentioned in the Performance docs https://deck.gl/docs/developer-guide/performance#supply-attributes-directly.

Does not seem so from your code sample. Supplying attributes will be a lot faster than using the getPosition callback, especially when your positions are already in a flat array.

**Comment by juscc** [comment]
> Does not seem so from your code sample. Supplying attributes will be a lot faster than using the getPosition callback, especially when your positions are already in a flat array.

Apologies, I've been trying all the different variations. Can you please see this example? Is this the correct way to supply attributes directly? Also I just did a quick test with the below with transitions and it just hangs for some reason.. I'll add a link to a video soon. 

> ```
> _renderLayers() {
	const length = data.length / 2;
	const points = new Float32Array(data);
	console.log(length, points);
	return new ScatterplotLayer({
		id: 'scatter-plot',
		data: {
			length: length,
			attributes: {
				getPosition: { value: points, size: 2 },
			},
		},
		radiusScale: 2,
		radiusMinPixels: 2,
		getRadius: 1,
		transitions: {
			getPosition: 500,
		},
		pickable: false,
	});
}
> ```

Thank you!

**Comment by Pessimistress** [comment]
Can you create a code sandbox? It's impossible to debug like this.

**Comment by juscc** [comment]
Of course sorry about that, here's the codesandbox I created https://codesandbox.io/s/react-hooks-scatterplot-transition-example-ocm1r?file=/src/vehicle-layer.js

Please comment out the transitions option in vehicle-layer.js to see the problem. Thank you!

**Comment by Pessimistress** [comment]
Some observations:

- You are updating data on every animation frame, so the transitions never gets a chance to evaluate.
- You have too many `useEffect` that depends on each other - this means multiple React rerenders on every animation frame, likely causing the performance issues

**Comment by juscc** [comment]
Thanks for that, I've reduced the useEffect hooks and it's performing much better.
I've successfully had the transition interpolation work when using  
```
 getPosition: (object, { index, data, target }) => {
      target[0] = data.src[index * DATA_POINT_LENGTH]; // long
      target[1] = data.src[index * DATA_POINT_LENGTH + 1]; // lat
      return target;
    },
```
but not with the below, should transitions work when supplying the data attributes directly? 
```
     data: {
       length: length,
       attributes: {
         getPosition: { value: points, size: 2 }
       }
     },
```

I've updated my codesandbox with both examples, you can uncomment the supply attributes directly approach to see the problem https://codesandbox.io/s/react-hooks-scatterplot-transition-example-ocm1r?file=/src/vehicle-layer.js

**Comment by Pessimistress** [comment]
Looks like there is a bug in the transition system. If you change `Float32Array` to `Float64Array` it will work.

**Comment by Pessimistress** [comment]
Using `Float32Array` for `getPosition` and transition is fixed in 8.3.8.
